From: Liam Howlett <liam.howlett@oracle.com>
Subject: mm/mmap: allow vma_expand() to lock both anon and file locks
Date: Thu, 16 Jun 2022 18:33:06 +0000

vma_expand() was only locking either file or anon based locks but could
unlock both.  This caused a locking error in trinity fuzzer.  Allow both
locks to be taken and released.

Link: https://lkml.kernel.org/r/20220616183258.1153609-1-Liam.Howlett@oracle.com
Fixes: a5985220056d (mm/mmap: use advanced maple tree API for mmap_region())
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/mmap.c |   17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

--- a/mm/mmap.c~mm-mmap-use-advanced-maple-tree-api-for-mmap_region-fix-3
+++ a/mm/mmap.c
@@ -694,6 +694,11 @@ inline int vma_expand(struct ma_state *m
 
 	vma_adjust_trans_huge(vma, start, end, 0);
 
+	if (anon_vma) {
+		anon_vma_lock_write(anon_vma);
+		anon_vma_interval_tree_pre_update_vma(vma);
+	}
+
 	if (file) {
 		mapping = file->f_mapping;
 		root = &mapping->i_mmap;
@@ -701,9 +706,6 @@ inline int vma_expand(struct ma_state *m
 		i_mmap_lock_write(mapping);
 		flush_dcache_mmap_lock(mapping);
 		vma_interval_tree_remove(vma, root);
-	} else if (anon_vma) {
-		anon_vma_lock_write(anon_vma);
-		anon_vma_interval_tree_pre_update_vma(vma);
 	}
 
 	vma->vm_start = start;
@@ -732,15 +734,16 @@ inline int vma_expand(struct ma_state *m
 		mm->highest_vm_end = vm_end_gap(vma);
 	}
 
+	if (file) {
+		i_mmap_unlock_write(mapping);
+		uprobe_mmap(vma);
+	}
+
 	if (anon_vma) {
 		anon_vma_interval_tree_post_update_vma(vma);
 		anon_vma_unlock_write(anon_vma);
 	}
 
-	if (file) {
-		i_mmap_unlock_write(mapping);
-		uprobe_mmap(vma);
-	}
 
 	if (remove_next) {
 		if (file) {
_
